lib/blob/blobstore.c

```diff
@@ -2798,6 +2798,14 @@ blob_ancestor_calc_lba_and_lba_count(struct spdk_blob *blob, uint64_t io_unit, u
 	}
 	*lba_count = length;
 	while (blob->parent_id != SPDK_BLOBID_INVALID) {
+		uint32_t cluster_start_page = bs_io_unit_to_cluster_start(blob, io_unit);
+		bool is_valid_range = blob->back_bs_dev->is_range_valid(blob->back_bs_dev,
+				      bs_dev_page_to_lba(blob->back_bs_dev, cluster_start_page),
+				      bs_dev_byte_to_lba(blob->back_bs_dev, blob->bs->cluster_sz));
+		if (!is_valid_range) {
+			goto error;
```

Comment:
This happens when there's been a resize? If so why do we go to error in this case?

Reply:
> This happens when there's been a resize? If so why do we go to error in this case?

Yes we'd land in this case if resize happened. Can simply do a `return false` also from here, which would be more clear I think.

Reply:
hmm but this means we won't take into account the previous snapshots on the used space?

Reply:
> hmm but this means we won't take into account the previous snapshots on the used space?

Yes, if there is any snapshot further up the chain, we don't consider a read valid even if it contains that range. Jim confirmed that on original change.

```
Let's say we have blob A which is 10GB. We clone to get blob B and shrink it to 5GB. Then we clone again to get blob C and grow it to 8GB.

If we reference offset 5GB-8GB it is not valid to access blob A for those clusters! When we shrunk blob B, we automatically made those invalid.
```

Reply:
but this function is not about what's readable, it's to calculate how much data we have in the tree

Reply:
No, this function is called during rebuild if an allocated cluster is found, to check whether the snapshots have it allocated. @hrudaya21 ?

Reply:
Ah sorry the naming confused me :D